Sources/Workspace/Workspace.swift

```diff
@@ -252,24 +252,44 @@ public class Workspace {
         pinsFile: AbsolutePath,
         manifestLoader: ManifestLoaderProtocol,
         repositoryManager: RepositoryManager? = nil,
-        currentToolsVersion: ToolsVersion = ToolsVersion.currentToolsVersion,
-        toolsVersionLoader: ToolsVersionLoaderProtocol = ToolsVersionLoader(),
+        currentToolsVersion: ToolsVersion? = nil,
+        toolsVersionLoader: ToolsVersionLoaderProtocol? = nil,
         delegate: WorkspaceDelegate? = nil,
-        config: Workspace.Configuration = Workspace.Configuration(),
-        fileSystem: FileSystem = localFileSystem,
-        repositoryProvider: RepositoryProvider = GitRepositoryProvider(),
+        config: Workspace.Configuration? = nil,
+        fileSystem: FileSystem? = nil,
+        repositoryProvider: RepositoryProvider? = nil,
         identityResolver: IdentityResolver? = nil,
-        httpClient: HTTPClient = HTTPClient(),
+        httpClient: HTTPClient? = nil,
         netrcFilePath: AbsolutePath? = nil,
-        archiver: Archiver = ZipArchiver(),
-        checksumAlgorithm: HashAlgorithm = SHA256(),
-        additionalFileRules: [FileRuleDescription] = [],
-        isResolverPrefetchingEnabled: Bool = false,
-        enablePubgrubResolver: Bool = false,
-        skipUpdate: Bool = false,
-        enableResolverTrace: Bool = false,
+        archiver: Archiver? = nil,
+        checksumAlgorithm: HashAlgorithm? = nil,
+        additionalFileRules: [FileRuleDescription]? = nil,
+        isResolverPrefetchingEnabled: Bool? = nil,
+        enablePubgrubResolver: Bool? = nil,
+        skipUpdate: Bool? = nil,
+        enableResolverTrace: Bool? = nil,
```

Comment:
Since we're moving to instance APIs, can we add a diagnostic engine to the initializer instead of passing one to individual calls?

Reply:
Follow-up PR sounds good to me.